test/conversions.jl

```diff
@@ -114,29 +182,232 @@ end
+    @test promote_type(ARGB32, HSVA{Float16}) == AlphaColor{C,Float32,4} where {C<:Color3{Float32}}
+    @test promote_type(ARGB32, AHSV{Float16}) == AlphaColor{C,Float32,4} where {C<:Color3{Float32}}
+
+    @test promote_type(RGB, HSV) == Color3
+    @test promote_type(RGB, HSV{Float16}) == Color3
+    @test promote_type(HSV, RGB{Float16}) == Color3
+
+    @test promote_type(RGB, HSVA) == ColorAlpha{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(AHSV, RGB) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(RGB, HSVA{Float16}) == ColorAlpha{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(AHSV, RGB{Float16}) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+
+    @test promote_type(HSV, RGBA) == ColorAlpha{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(ARGB, HSV) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(HSV, RGBA{Float16}) == ColorAlpha{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(ARGB, HSV{Float16}) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+
+    @test promote_type(HSVA, RGBA) == ColorAlpha{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(ARGB, AHSV) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(HSVA, RGBA{Float16}) == ColorAlpha{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(ARGB, AHSV{Float16}) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+
+    @test promote_type(AHSV, RGBA) == AlphaColor{C,T,4} where {T, C<:Color3{T}} # != Transparent3
+    @test promote_type(ARGB, HSVA) == AlphaColor{C,T,4} where {T, C<:Color3{T}} # != Transparent3
+    @test promote_type(AHSV, RGBA{Float16}) == AlphaColor{C,T,4} where {T, C<:Color3{T}} # != Transparent3
+    @test promote_type(ARGB, HSVA{Float16}) == AlphaColor{C,T,4} where {T, C<:Color3{T}} # != Transparent3
+
+    @test promote_type(RGB24, HSV) == Color3
+    @test promote_type(RGB24, HSVA{Float16}) == ColorAlpha{C,Float32,4} where C<:Color3{Float32}
+    @test promote_type(ARGB32, HSV) == AlphaColor{C,T,4} where {T, C<:Color3{T}}
+    @test promote_type(ARGB32, HSVA{Float16}) == AlphaColor{C,Float32,4} where C<:Color3{Float32}
+end
+
+@testset "promotions with abstract types" begin
+    @test promote_type(Colorant, RGB{Float64}) == AlphaColor{C,T} where {T, C<:Color{T}}
+    @test promote_type(Colorant{N0f8}, RGB{Float64}) == AlphaColor{C,Float64} where {C<:Color{Float64}}
+    @test promote_type(Colorant{Float32,3}, RGB{Float64}) == AlphaColor{C,Float64,4} where C<:Color{Float64,3}
+    @test promote_type(Color, RGB{Float64}) === Color
+    @test promote_type(Color{N0f8}, RGB{Float64}) === Color{Float64}
+    @test promote_type(Color{Float32,3}, RGB{Float64}) === Color{Float64,3}
+    @test promote_type(AbstractRGB, RGB{Float64}) === RGB
+    @test promote_type(AbstractRGB{N0f8}, RGB{Float64}) === RGB{Float64}
+    @test promote_type(TransparentColor, RGB{Float64}) == AlphaColor{C,T} where {T, C<:Color{T}}
+    @test promote_type(TransparentColor{RGB{N0f8}}, RGB{Float64}) === ARGB
```

Comment:
Argh, this is indeed why it might be nice to have the `T` constrained to `eltype(C)`.